### PR TITLE
feat: display per-device sampling frequency in session legend

### LINKED ISSUE #199: feat: Display sample frequency information in session UI
## Summary
Add sample frequency information above the session channel and device serial number to provide better context about the data collection rate.

## Current Behavior
Users cannot see the sampling frequency used for a logged session without examining the data or remembering the settings used during collection.

## Proposed Enhancement
Display the sampling frequency prominently above the session channel and device serial number information in the UI.

## Example Display
```
100 Hz
Channel: AI0
Device: DAQ-12345
```

## Benefits
- Quick identification of sampling rate without opening session
- Better understanding of data resolution
- Helps users identify sessions collected at specific rates

## Acceptance Criteria
- [ ] Sampling frequency is displayed above channel information
- [ ] Format is clear and consistent (e.g., "100 Hz", "1 kHz")
- [ ] Works for single and multi-device sessions
- [ ] Handles variable sampling rates appropriately

## Labels
enhancement, ui

🤖 Generated with [Claude Code](https://claude.ai/code)

### PR DESCRIPTION
**Pull request by tylerkron**
## Summary
Closes #199. Adds per-device sampling frequency display to the session legend panel, captured at the start of each logging session.

- New `SessionDeviceMetadata` table (composite PK on `LoggingSessionID` + `DeviceSerialNo`) stores the configured `SamplingFrequencyHz`, `DeviceName`, and serial for every device that owns at least one subscribed channel when streaming starts.
- `DeviceLegendGroup` gains `SamplingFrequencyHz` + a `FormatFrequency` helper that renders `100 Hz` / `1 kHz` / `1.5 kHz` / `2 MHz`. The legend XAML shows it above the device serial; the row is hidden via `BoolToVis` for legacy sessions that have no metadata.
- `DatabaseLogger.DisplayLoggingSession` loads metadata on the background thread and swaps it onto the UI thread atomically with the rest of the session state.
- Cascade deletion is wired through every path: EF cascade for `Sessions.Remove` / `RemoveRange`, plus explicit `DELETE FROM SessionDeviceMetadata` in both raw-SQL deletion sites (`LoggingManager.DeleteLoggingSessionIfPresent`, `DatabaseLogger.DeleteLoggingSession`). The "delete all" path nukes the DB file so it's covered automatically.
- Multi-device sessions naturally produce one metadata row per device, so each device shows its own rate.
- Defensive dedupe by serial when iterating `ConnectedDevices` — a transient duplicate during USB re-enumeration would otherwise hit the composite-PK constraint and abort the session start.

This same table is well-positioned to also hold per-device sample counts when #198 is implemented.

## Test plan
- [ ] Start a single-device streaming session and confirm the legend shows the configured frequency above the device serial
- [ ] Start a multi-device session with each device at a different rate and confirm each legend group shows its own frequency
- [ ] Open a session that was logged before this change and confirm the frequency row is hidden (no layout artifacts)
- [ ] Delete a logging session via the per-row delete and confirm the corresponding `SessionDeviceMetadata` rows are gone
- [ ] Delete all sessions and confirm the database recreates cleanly
- [ ] Run `dotnet test Daqifi.Desktop.Test` on Windows — new tests:
  - `DeviceLegendGroupTests` (11 cases) — `FormatFrequency` edge cases, property-change propagation, truncated serial
  - `SessionDeviceMetadataTests` (3 cases) — cascade delete via EF, multi-device per session, composite-key uniqueness

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: Daqifi.Desktop/Loggers/LoggingManager.cs [hunk 1]
```diff
@@ -404,6 +414,59 @@ public void Unsubscribe(IChannel channel)
...
+            }
+
+            yield return new SessionDeviceMetadata
+            {
+                LoggingSessionID = sessionId,
+                DeviceSerialNo = device.DeviceSerialNo,
+                DeviceName = device.Name ?? string.Empty,
+                SamplingFrequencyHz = device.StreamingFrequency
+            };
+        }
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply Daqifi.Desktop/Loggers/LoggingManager.cs]: **Fixed in 300456d.**

Refactored `BuildDeviceMetadataForSession` to:
- Snapshot `SubscribedChannels` and `ConnectionManager.Instance.ConnectedDevices` to local lists upfront, so iteration is never against the live mutable collections.
- Return a fully materialized `List<SessionDeviceMetadata>` instead of `yield return`, so the caller is never lazily enumerating during DB operations.
- Dedupe by serial number with a `HashSet<string>(StringComparer.Ordinal)` (serials are opaque identifiers, not culture-sensitive text).
- Additionally wrapped the metadata-capture call in `OnActiveChanged` in a try/catch so the session still starts even if metadata enumeration fails (logged and skipped).

**Comment by tylerkron** [comment]
## Qodo summary review — responses

Addressing the items from Qodo's review summary that weren't posted as inline comments:

### 📘 Rule violations

**1. Test `Cleanup()` swallows exceptions silently** — *Disagree, no change.*
This matches the existing pattern in `ExportPerformanceTests` for best-effort temp-file cleanup. SQLite file deletion can fail transiently on Windows when the connection pool hasn't fully released the file handle; throwing from `[TestCleanup]` would mask the actual test result. The `SqliteConnection.ClearAllPools()` call already handles the common case.

**2. `LoadSessionDeviceFrequency` uses sync EF on what could be a background thread** — *Disagree, no change.*
The surrounding `DisplayLoggingSession` code is entirely synchronous EF running on a background thread (kicked off from `Task.Run` in the caller). Mixing one async EF call into an otherwise-sync method would be inconsistent and offer no benefit — the thread is already off the UI. Converting the whole load path to async is a separate refactor and out of scope here.

**3. Serial number comparisons lack normalization** — *Partially fixed in 300456d.*
Added `StringComparer.Ordinal` for the dedup `HashSet`s in `BuildDeviceMetadataForSession` (serials are opaque identifiers — ordinal is the correct semantic). I deliberately did **not** add `Trim()`: `DataSample.DeviceSerialNo` is stored without trimming elsewhere in the codebase, so trimming only on the metadata path would desync the join key between metadata and samples.

**4. Overlong fluent API line in `LoggingContext.cs`** — *Fixed in 300456d.* Wrapped across multiple lines.

### 🐞 Bug

**Unsafe list enumeration in `BuildDeviceMetadataForSession`** — *Fixed in 300456d.* See [reply on the inline comment](https://github.com/daqifi/daqifi-desktop/pull/476#discussion_r3083225464).

File: Daqifi.Desktop/Loggers/LoggingManager.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tylerkron) [reply Daqifi.Desktop/Loggers/LoggingManager.cs]: Already addressed in commit 300456d — `BuildDeviceMetadataForSession` now snapshots `SubscribedChannels` (and the device's channels) into a local list before enumeration, eliminating the mutate-during-enumeration risk.

File: Daqifi.Desktop/MainWindow.xaml [hunk 2]
```diff
@@ -751,7 +815,36 @@
...
+                                                                                           FontSize="8"
+                                                                                           Opacity="0.5"
+                                                                                           VerticalAlignment="Center"/>
+                                                                                <TextBlock Text="{Binding SampleCountDisplay}"
+                                                                                           FontSize="10"
+                                                                                           Opacity="0.85"
+                                                                                           VerticalAlignment="Center"/>
+                                                                            </StackPanel>
+                                                                        </StackPanel>
+                                                                    </StackPanel>
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply Daqifi.Desktop/MainWindow.xaml:849]: Disagreeing with this one. We don't currently track total byte size anywhere — sample count was the additional metric the user explicitly requested in this iteration. Adding a byte-size column would require a non-trivial schema change (per-row size accounting or an aggregate column wired through DatabaseLogger and SdCardSessionImporter), which is out of scope for this PR. Happy to break it out into a follow-up issue if the user wants it.

File: Daqifi.Desktop/Loggers/LoggingSession.cs [hunk 3]
```diff
@@ -8,20 +8,105 @@ public class LoggingSession : ObservableObject
...
+    public bool HasMultipleDevices => DeviceCount > 1;
+
+    /// <summary>
+    /// Sample count formatted with thousands separators (e.g. "16,000"),
+    /// or empty string when the count has not been recorded yet.
+    /// </summary>
+    [NotMapped]
+    public string SampleCountDisplay => SampleCount.HasValue
+        ? SampleCount.Value.ToString("N0", System.Globalization.CultureInfo.CurrentCulture)
+        : string.Empty;
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply Daqifi.Desktop/Loggers/LoggingSession.cs:108]: Fixed in abf4413. `SampleCountDisplay` now renders compact magnitude notation (e.g. `1.23M`, `16K`, `2.4B`) via `LoggingSession.FormatAbbreviated`. The full precise count with thousands separators is preserved as `SampleCountTooltip` and surfaced via `ToolTip` on the row and chart-header SAMPLES segments. The chart header uses a parallel `CurrentSessionSampleCountDisplay` on `DatabaseLogger` so live and persisted views stay visually consistent.

File: Daqifi.Desktop/MainWindow.xaml [hunk 4]
```diff
@@ -654,9 +717,9 @@
                                             <ItemsControl ItemsSource="{Binding DbLogger.DeviceLegendGroups}">
                                                 <ItemsControl.ItemTemplate>
                                                     <DataTemplate DataType="{x:Type logger:DeviceLegendGroup}">
-                                                        <StackPanel Margin="0,2,0,4">
-                                                            <!-- Device Header -->
-                                                            <TextBlock FontSize="9" Opacity="0.6" Margin="0,0,0,2"
+                                                        <StackPanel Margin="0,4,0,8">
+                                                            <!-- Device Header (sampling frequency now lives in the chart subtitle) -->
+                                                            <TextBlock FontSize="10" Opacity="0.7" Margin="0,0,0,3"
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply Daqifi.Desktop/MainWindow.xaml:723]: Disagreeing — this was a deliberate UX change requested by the user. Earlier in the PR I had a frequency badge in the legend, but the user asked us to move it. The sampling frequency now lives in the centered chart header strip directly above the plot (`SAMPLE RATE  10 Hz` segment in the WPF header), and is also shown inline on every session list row. The legend was simplified to just the device serial because the duplication felt cluttered. The requirement to surface frequency "in the session UI" is met — just in the new header location instead of the legend.

File: Daqifi.Desktop/Loggers/LoggingManager.cs [hunk 5]
```diff
@@ -468,14 +572,87 @@ public ObservableCollection<LoggingSession> LoadPersistedLoggingSessions()
...
+
+            var tracked = context.Sessions.FirstOrDefault(s => s.ID == session.ID);
+            if (tracked != null)
+            {
+                tracked.SampleCount = count;
+                context.SaveChanges();
+            }
+            // Also surface immediately on the in-memory entity bound by the UI.
+            session.SampleCount = count;
+        }
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply Daqifi.Desktop/Loggers/LoggingManager.cs:627]: Real bug, fixed in abf4413. Added `DatabaseLogger.WaitForIdle(timeout)` which blocks until `_buffer.Count == 0` and a new `_consumerBusy` flag is clear (the flag is set inside the consumer's flush block and reset after `BulkInsert` commits, with reset on the exception paths too). `PersistSessionSampleCount` now resolves the `DatabaseLogger` from `Loggers` and calls `WaitForIdle(TimeSpan.FromSeconds(10))` before issuing the COUNT, so the persisted value reflects every buffered row.

File: Daqifi.Desktop/Loggers/LoggingManager.cs [hunk 6]
```diff
@@ -54,6 +55,15 @@ partial void OnActiveChanged(bool oldValue, bool newValue)
...
                     }
+
+                    // Finalize the session by recording its sample count so the
+                    // list view never has to count rows. Runs on a background
+                    // thread because COUNT(*) on a multi-million-row session
+                    // would otherwise block the UI thread that toggled Active.
+                    // The list row updates automatically when SampleCount is
+                    // set, since LoggingSession is INotifyPropertyChanged.
+                    var sessionToFinalize = Session;
+                    _ = Task.Run(() => PersistSessionSampleCount(sessionToFinalize));
```
[the remark this answers is not in the document]
Comment (tylerkron) [reply Daqifi.Desktop/Loggers/LoggingManager.cs:66]: Real bug, fixed in abf4413. Both call sites that mutate `LoggingSession.SampleCount` from background threads now marshal the assignment through `Application.Current.Dispatcher.Invoke` while keeping the DB write on the background thread:

- `LoggingManager.PersistSessionSampleCount` (stream-stop path, runs under `Task.Run`)
- `SdCardSessionImporter.ImportSessionAsync` (SD import path)

The DB `SaveChanges` still happens off the UI thread; only the bound entity mutation that raises `PropertyChanged` is dispatched.
